cpp/src/gandiva/ilike_holder.cc

```diff
@@ -0,0 +1,101 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/ilike_holder.h"
+
+#include <regex>
+#include "gandiva/node.h"
+#include "gandiva/regex_util.h"
+
+namespace gandiva {
+RE2::Options IlikeHolder::regex_op_ = RE2::Options();
+
+RE2 IlikeHolder::starts_with_regex_(R"((?i)(\w|\s)*\.\*)");
+RE2 IlikeHolder::ends_with_regex_(R"((?i)\.\*(\w|\s)*)");
+RE2 IlikeHolder::is_substr_regex_(R"((?i)\.\*(\w|\s)*\.\*)");
+
+// Short-circuit pattern matches for the following common sub cases :
+// - starts_with, ends_with and is_substr
+const FunctionNode IlikeHolder::TryOptimize(const FunctionNode& node) {
+  std::shared_ptr<IlikeHolder> holder;
+  auto status = Make(node, &holder);
+  if (status.ok()) {
+    std::string& pattern = holder->pattern_;
+    auto literal_type = node.children().at(1)->return_type();
+
+    if (RE2::FullMatch(pattern, starts_with_regex_)) {
```

Comment:
this seems incorrect, they should be case-insensitive

Reply:
not really, the pattern has (?i)(https://github.com/google/re2/issues/197) to turn case-insensitive on for the expression.

Reply:
this matching doesn't matter..actual starts_with, ends_with, sub_str are not case-insensitive

Reply:
Here we match on the pattern for starts_with, ends_with cases, and replace with them..which won't be case-insensitive

Reply:
Ok, now both like and ilike are using the same method.

Reply:
This is not yet fixed. You are still replacing with starts_with, ends_with which are case-sensitive. You need to disable TryOptimise for ILIKE

Reply:
Done, sorry I thought first that it had no difference.